crates/miden-lib/asm/note_scripts/MINT.masm

```diff
@@ -19,43 +27,102 @@ const.ERR_MINT_WRONG_NUMBER_OF_INPUTS="MINT script expects exactly 9 note inputs
 #! Inputs:  [ARGS, pad(12)]
 #! Outputs: [pad(16)]
 #!
-#! Note inputs are assumed to be as follows (in order):
-#! - RECIPIENT: The recipient account ID (4 elements)
-#! - Output note config (4 elements):
-#!   - execution_hint: Execution hint for the output note
-#!   - note_type: Type of the output note
-#!   - aux: Auxiliary data for the output note
-#!   - tag: Note tag for the output note
+#! Note inputs support two modes. Depending on the number of note inputs,
+#! a private or public note is created on consumption of the MINT note:
+#!
+#! Private mode (8 inputs) - creates a private note:
+#! - execution_hint: Execution hint for the output note
+#! - aux: Auxiliary data for the output note
+#! - tag: Note tag for the output note
 #! - amount: The amount to mint
+#! - RECIPIENT: The recipient digest (4 elements)
+#!
+#! Public mode (12+ inputs) - creates a public note with variable-length inputs:
+#! - execution_hint: Execution hint for the output note
+#! - aux: Auxiliary data for the output note
+#! - tag: Note tag for the output note
+#! - amount: The amount to mint
+#! - SCRIPT_ROOT: Script root of the output note (4 elements)
+#! - SERIAL_NUM: Serial number of the output note (4 elements)
+#! - [INPUTS]: Variable-length inputs for the output note (Vec<Felt>)
+#!             The number of output note inputs = num_mint_note_inputs - 12
 #!
 #! Panics if:
 #! - account does not expose distribute procedure.
-#! - the number of inputs is not exactly 9.
+#! - the number of inputs is not exactly 8 for private or less than 12 for public output notes.
 begin
     dropw
     # => [pad(16)]
 
     # Load note inputs into memory starting at address 0
     push.0 exec.active_note::get_inputs
-    # => [num_inputs, inputs_ptr, pad(16)]
+    # => [total_inputs, inputs_ptr, pad(16)]
 
-    # Verify we have the correct number of inputs
-    eq.MINT_NOTE_INPUTS_NUMBER assert.err=ERR_MINT_WRONG_NUMBER_OF_INPUTS drop
-    # => [pad(16)]
+    dup
+    # => [num_inputs, num_inputs, inputs_ptr, pad(16)]
+
+    u32assert2.err=ERR_MINT_WRONG_NUMBER_OF_INPUTS
+    u32gte.MINT_NOTE_MIN_NUM_INPUTS_PUBLIC
+    # => [is_public_output_note, total_inputs, inputs_ptr, pad(16)]
+
+    if.true
+        # public output note creation
+
+        movdn.5 drop
+        # => [EMPTY_WORD, total_inputs, pad(16)]
+
+        mem_loadw_be.4
+        # => [SCRIPT_ROOT, total_inputs, pad(12)]
+
+        padw mem_loadw_be.8
+        # => [SERIAL_NUM, SCRIPT_ROOT, total_inputs, pad(12)]
```

Comment:
If on line 71 above we do `movdn.9 drop`, the stack state after that line would be:
```
[EMPTY_WORD, EMPTY_WORD, total_inputs, pad(8)]
```
And then we can avoid doing a `padw` here.